src/index.spec.ts

```diff
@@ -1,12 +1,196 @@
 import {
-  isInteger,
-  toLowerCase,
-  removeDuplicatesFromArray,
+  createProduct,
   createRandomProduct,
   getStarWarsPlanets,
-  createProduct 
+  isInteger,
+  removeDuplicatesFromArray,
+  toLowerCase,
 } from './index';
+import { faker } from '@faker-js/faker';
+import { createTestProduct } from './test/util';
+import users from './utils/users';
+import fetch from 'node-fetch';
+
+const { Response } = jest.requireActual('node-fetch');
+jest.mock('node-fetch');
+
+beforeEach(() => {
+  jest.resetAllMocks();
+});
+
+describe('Integer validation', () => {
+  const integerValues = [[1], [-1], [12451], [34566]];
+  test.each(integerValues)('returns true with integer %p', (value) => {
+    expect(isInteger(value)).toBe(true);
+  });
+
+  const stringValues = [['1'], ['-1'], ['ab'], ['123.3']];
+  test.each(stringValues)('returns false with string value %p', (value) => {
+    expect(isInteger(value)).toBe(false);
+  });
+
+  const decimalValues = [[12.3], [-3.4], [0.999]];
+  test.each(decimalValues)('returns false with decimal %p', (value) => {
+    expect(isInteger(value)).toBe(false);
+  });
+});
+
+describe('String to lowercase', () => {
+  test('fails with empty string', () => {
+    expect(toLowerCase('')).toMatchInlineSnapshot(`"Please provide a string"`);
+  });
+
+  const validStringCases = [
+    ['HELLO', 'hello'],
+    ['ByE', 'bye'],
+    ['123', '123'],
+    ['abc!de', 'abc!de'],
+  ];
+  test.each(validStringCases)(
+    'works with valid string value %p',
+    (value: string, expected: string) => {
+      expect(toLowerCase(value)).toBe(expected);
+    },
+  );
+});
+
+describe('Remove duplicates from array', () => {
+  test('returns empty array when an empty array is received', () => {
+    const result = removeDuplicatesFromArray([]);
+    expect(result).toEqual([]);
+  });
+
+  const intArrayCases = [
+    [[1], [1]],
+    [[1, 1], [1]],
+    [
+      [1, 2, 2, 3, 3],
+      [1, 2, 3],
+    ],
+  ];
+  test.each(intArrayCases)('works with int array %p', (array, expected) => {
+    expect(removeDuplicatesFromArray(array)).toEqual(expected);
+  });
+
+  const stringArrayCases = [
+    [['a'], ['a']],
+    [['a', 'a'], ['a']],
+    [
+      ['a', 'b', 'b', 'c', 'c'],
+      ['a', 'b', 'c'],
+    ],
+  ];
+  test.each(stringArrayCases)(
+    'works with string array %p',
+    (array, expected) => {
+      expect(removeDuplicatesFromArray(array)).toEqual(expected);
+    },
+  );
+});
+
+describe('create product', () => {
+  test('returns the product with its id replaced', () => {
+    const product = createTestProduct();
+    const updated = createProduct(product);
+
+    expect(product.id).not.toBe(updated.id);
+  });
+
+  const productNameOverrides = [[''], ['!%!'], [faker.random.alphaNumeric(31)]];
+  test.each(productNameOverrides)(
```

Comment:
This could also be simplified to a single test, but I do understand why you would test it like this

Reply:
I mean, you could have replaced many properties at once. Eg. name, description, tags